keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1218 @@
+
+#### Google ConfigSync
+
+**Pattern**: list of Group-Version-Kind-Namespace-Name (GVKNN) (from ResourceGroup)
+
+Distinct sets of synchronized resources are represented by RootSync / RepoSyncs, along with a ResourceGroup that has the full inventory.  Each object has some annotations that define membership, including the same `config.k8s.io/owning-inventory` as is used by kpt.    As with other solutions, following the “chain” from RootSync/RepoSync to managed objects is somewhat obscure.
+
+
+## Proposal
+
+A v2-prunable "apply set" is associated with an object on the cluster.   We define a set of standardized labels and annotations that identify the “parent object” of the apply set and the “member objects” of that parent.  We operate at the plumbing layer; we aim to support:
+
+- listing the parent objects efficiently (porcelain may expose this as listing groups of objects as managed by the various tools)
+- listing the member objects for a specific parent object efficiently (porcelain may use this for advanced diffing and pruning, and for presenting objects grouped by their higher-level aggregation)
+- basic apply-with-prune operations, where it creates or reuses a Secret in the cluster as the parent object.
+
+### Apply Set
+
+"Apply set" refers to a group of resources that are applied to the cluster by a tool. An apply set has a “parent” object of the tool’s preference. This “parent” object can be implemented using a ConfigMap, Secret, or a CRD of the tool’s choice.
+
+“ApplySet” is used to refer to the parent object in this design document, though the actual concrete resource on the cluster will typically be of a different Kind.  We might think of ApplySet as a “duck-type” based on the `applyset.k8s.io/id` label proposed here.
+
+### Member Object Labels
+
+Objects that are part of an ApplySet should carry a standardized label, with a key of:
+
+```yaml
+applyset.k8s.io/part-of: <applysetkey>
+```
+
+The `<applysetkey>` can be chosen essentially arbitrarily (subject to the limits of label values).
+
+We recommend that tooling uses a prefix that is its "tool name", e.g. `kubectl.` or `helm.`,
+followed by some tooling-specific unique value.  We can investigate maintaining a formal registry,
+but we initially reserve the following prefixes:
+
+* `kubectl.` for kubectl
+* `helm.` for helm
+* `kpt.` for kpt
+* `uid.` for where the suffix is the UID of the applyset object (e.g. `uid.e049464e-4583-4642-9649-93dcb0e96bd4`)
+* `id.` for where the suffix is the group-kind followed by the namespace followed by the name, such as
+  `id.configmaps.ns1.parent1`.  (While this is a little tricky to parse, it should be unique because
```

Comment:
Alternative to my proposal above to not dictate the format, we could _require_ that it be something specific, or _require_ that this data be embedded in a second label. Why? To mitigate over-selection risk. Here's my train of thought, which could perhaps go in the "downsides" section if we keep the design as-is:

There is still a risk of under- and over- selection in the new implementation. Under-selection will only occur when manual changes are made out of band, which we can document as not supported. I believe over-selection is the bigger risk, because although it will now be possible (and encouraged) for each set to get a scope-unique or even globally unique identifier, we have no possible way of enforcing this. Users could create multiple applyset parent objects that use the same applyset name and have overlapping GK lists within the same scope (e.g. namespace). This scenario would be undetectable by our tooling, and if the apply operations are associated with two different sets of objects, the associated pruning calculations will scope in each other's objects, with unexpected and undesirable results. 

This scenario strikes me as highly analogous to failure mode of controllers with overlapping selectors in the days before `ownerReferences.controller` was introduced. Is an analogous solution available? Perhaps instead of suggesting that applyset names embed metadata about the parent, we could dictate their names, or have a second, required label that identifies the "controller" in a manner that compliant tools MUST validate before mutating the set. Using UID and/or GVKNN there would make it analogous to controllerRef (we could even literally serialize one?), and we could make the value opaque (e.g. base64 encoded?) to discourage (though in no way prevent) manual edits. Of course, especially with no server-side component, we have no way whatsoever of enforcing this, and would be reliant on correct implementation by tools opting in to this standard. Still, that's got to be a lot safer than relying on end users of the command.

Thoughts?

Reply:
I do think we can detect if there is a conflicting "parent", though we have to query for all the Secrets/ConfigMaps/labeled-CRDs first.  I think what's important is that it should be opaque, i.e. we shouldn't rely on it having a certain value for selection to work.  But opaque doesn't mean that the value has to be chosen randomly; I do think that both UUID and base64(gvknn) would work well.

I do think we'll be able to understand the tradeoffs better here if we can move to implementation.  We'll likely find cheaper ways to do some of the operations, and likely a few validations that we've overlooked :-)